### ISSUE TITLE
Sending SIGTERM to Postgres causes infinite shutdown

### ISSUE THREAD

**Issue by sdudoladov** [zalando/patroni#1175]
Sending `SIGTERM` directly to the root postmaster process puts the PG cluster into infinite shutting down state while retaining the leader lock.

To reproduce, `kill`  the ` /usr/lib/postgresql/11/bin/postgres -D /home/postgres/pgdata ...` process.

Sample Patroni log:
```
INFO: Lock owner: postgresql0; I am postgresql0
INFO: no action.  i am the leader with the lock
CEST [21150] LOG:  received fast shutdown request
CEST [21150] LOG:  aborting any active transactions
CEST [21162] FATAL:  terminating connection due to administrator command
CEST [21150] LOG:  worker process: logical replication launcher (PID 21160) exited with exit code 1
CEST [21155] LOG:  shutting down
CEST [21150] LOG:  database system is shut down
INFO: Lock owner: postgresql0; I am postgresql0
...
```
in Postgres log 
```
FATAL,57P03,"the database system is shutting down",,,,,,,,,""
FATAL,57P03,"the database system is shutting down",,,,,,,,,""
...
```

`ps -ef` within a Spilo container after kill
```
...
postgres    74    72  0 Sep10 ?        00:00:02 postgres: acid-test-cluster: logger   
postgres    79    72  0 Sep10 ?        00:00:00 postgres: acid-test-cluster: checkpointer   
postgres    83    72  0 Sep10 ?        00:00:00 postgres: acid-test-cluster: archiver   last was 00000
postgres    84    72  0 Sep10 ?        00:00:01 postgres: acid-test-cluster: stats collector   
postgres    93    72  0 Sep10 ?        00:00:13 postgres: acid-test-cluster: postgres postgres [local] 
```
this last process with `PID` 93 likely prevents [smart shutdown](https://www.postgresql.org/docs/10/server-shutdown.html)

**Comment by sdudoladov** [comment]
For sure, but it is still a corner case - albeit an improbable one -  where the leader lock is held by an unhealthy node, and an automatic failover can remedy the situation.

**Comment by CyberDem0n** [comment]
Patroni is waiting for postgres shutdown and therefore maintaining a leader lock.
This is by design.
Sigterm is an equivalent of smart shutdown. Postgres will wait until existing connections are closed, and will not close them forcibly. Existing connections can still run transactions. If you promote the replica in such situation you will create a split-brain. We don't want that to happen.

**Comment by sdudoladov** [comment]
Sorry, I just realized I did not write the original description precisely enough

TL;DR it is the connection from Patroni itself that prevents smart shutdown

Steps to reproduce locally:
1. Start Patroni
```
$ etcd --data-dir=data/etcd
$ ./patroni.py postgres0.yml
```
2. Send sigterm to postgres. That will leave the only connection open visible with `ps -ef | grep [p]ostgres`
```
29998 29988  0 16:51 ?        00:00:00 postgres: batman: postgres postgres 127.0.0.1(57838) idle
```
With Spilo the same line reads
```
postgres  1561  1405  0 14:30 ?        00:00:00 postgres: batman-cluster: postgres postgres [local]
```

Patroni log at that point:
```
2019-09-13 16:46:19.743 CEST [29550] LOG:  received smart shutdown request
2019-09-13 16:46:19.748 CEST [29550] LOG:  worker process: logical replication launcher (PID 29561) exited with exit code 1
2019-09-13 16:46:21,076 INFO: Lock owner: postgresql0; I am postgresql0
2019-09-13 16:46:21,088 INFO: no action.  i am the leader with the lock
2019-09-13 16:46:31,078 INFO: Lock owner: postgresql0; I am postgresql0
2019-09-13 16:46:31,093 INFO: no action.  i am the leader with the lock
2019-09-13 16:46:41,075 INFO: Lock owner: postgresql0; I am postgresql0
```
3.  Terminating the remaining connection process directly manifests in logs as follows
```
2019-09-13 16:51:01,085 INFO: no action.  i am the leader with the lock
2019-09-13 16:51:05.665 CEST [29564] FATAL:  terminating connection due to administrator command
2019-09-13 16:51:05.670 CEST [29556] LOG:  shutting down
2019-09-13 16:51:05.681 CEST [29550] LOG:  database system is shut down
2019-09-13 16:51:11,075 WARNING: Postgresql is not running.
2019-09-13 16:51:11,076 INFO: Lock owner: postgresql0; I am postgresql0
```
4. After that Patroni manages the recovery perfectly by restarting the PG server.
```
2019-09-13 16:51:12,243 INFO: Lock owner: postgresql0; I am postgresql0
2019-09-13 16:51:12,244 INFO: establishing a new patroni connection to the postgres cluster
2019-09-13 16:51:12,272 INFO: promoted self to leader because i had the session lock
server promoting
2019-09-13 16:51:12.284 CEST [29990] LOG:  received promote request
2019-09-13 16:51:12.284 CEST [29990] LOG:  redo is not required
2019-09-13 16:51:12,284 INFO: cleared rewind state after becoming the leader
2019-09-13 16:51:12.288 CEST [29990] LOG:  selected new timeline ID: 2
2019-09-13 16:51:12.343 CEST [29990] LOG:  archive recovery complete
2019-09-13 16:51:12.350 CEST [29988] LOG:  database system is ready to accept connections
```

**Comment by sdudoladov** [comment]
Doing the same on the replica tricks Patroni into thinking it has a healthy replica while in fact the replica is shutting down.
Running `/usr/lib/postgresql/11/bin/pg_ctl stop -m s` seems to have the same effect but I did not thoroughly checked it.

**Comment by CyberDem0n** [comment]
> Sorry, I just realized I did not write the original description precisely enough

I perfectly know what you wanted to tell.

For your specific case it just happened that there was only one connection to the cluster (from Patroni). In fact there might be also connections from the applications and during the smart shutdown postgres will wait until all connections are closed. Open connections doesn't get any notifications that the cluster is being shutdown. This is the _intentional_ behavior of postgres.

Patroni itself is never doing the smart shutdown. Since you did it **manually**, you basically shoot you own foot, therefore you should not complain that it hurts!

**Comment by sdudoladov** [comment]
now I see what you mean. Since this is not a real issue, let's not discuss it any further
